Rework the printing of attributes

vgvassilev opened the pull request:
Commit https://github.com/llvm/llvm-project/commit/46f3ade introduced a notion of printing the attributes on the left to improve the printing of attributes attached to variable declarations. The intent was to produce more GCC compatible code because clang tends to print the attributes on the right hand side which is not accepted by gcc.

This approach has increased the complexity in tablegen and the attrubutes themselves as now the are supposed to know where they could appear. That lead to mishandling of the `override` keyword which is modelled as an attribute in clang.

This patch takes an inspiration from the existing approach and tries to keep the position of the attributes as they were written. To do so we use simpler heuristic which checks if the source locations of the attribute precedes the declaration. If so, it is considered to be printed before the declaration.

Fixes https://github.com/llvm/llvm-project/issues/87151

vgvassilev committed:
Revert "Remove extra switch from  0323938d"

This reverts commit 6c0b9e35761738726aded3b399db89eb0a86f82b.

vgvassilev committed:
Revert "Fix warning in MSVC"

This reverts commit 0323938d3c7a1a48b60a7dea8ec7300e98b4a931.

vgvassilev committed:
Revert "Fix ast print of variables with attributes"

This reverts commit 46f3ade5083b8bfce55c78a21086a487eaac6f99.

vgvassilev committed:
Rework the printing of attributes.

Commit llvm/llvm-project@46f3ade introduced a notion of printing the attributes
on the left to improve the printing of attributes attached to variable
declarations. The intent was to produce more GCC compatible code because clang
tends to print the attributes on the right hand side which is not accepted by
gcc.

This approach has increased the complexity in tablegen and the attrubutes
themselves as now the are supposed to know where they could appear. That lead to
mishandling of the `override` keyword which is modelled as an attribute in clang.

This patch takes an inspiration from the existing approach and tries to keep
the position of the attributes as they were written. To do so we use simpler
heuristic which checks if the source locations of the attribute precedes the
declaration. If so, it is considered to be printed before the declaration.

Fixes llvm/llvm-project#87151

vgvassilev committed:
OpenMP attributes now respect the pragma location

vgvassilev commented:
@giulianobelinassi, I could not select you for a reviewer but please take a look at the pull request.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -129,11 +118,13 @@ namespace {
-    }
-
+    enum class AttrPosAsWritten {
+      Unknown = 0,
```
'Unknown' is likely to happen with 'implicit' attributes.  Perhaps the 'CanPrintOnLeft' is worth keeping so that we can aggressively print those on the left?

vgvassilev replied:
Unknown means that we have no preference and we want to keep the old behavior of `prettyPrintAttributes`.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +241,43 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
 
-  return canPrintOnLeftSide(A->getKind());
-}
+  if (C.getSourceManager().isBeforeInTranslationUnit(ALoc, DLoc))
```
I'm not particularly up-to-date on source-location comparisons, but I hope you have a test to make sure that this works for an attribute defined in a macro, but put on the RHS.

vgvassilev replied:
Added a test. Was that the case you had in mind?

erichkeane replied:
Yep, pretty much.  I would also like to see a DIFFERENT attribute on each side (perhaps multiples on each side?) to make sure we properly maintain ordering, but the one you have covers most of the concern

vgvassilev replied:
Test along these lines was added.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +241,43 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
-void DeclPrinter::prettyPrintAttributes(Decl *D, llvm::raw_ostream &Out,
-                                        AttrPrintLoc Loc) {
+void DeclPrinter::prettyPrintAttributes(const Decl *D,
+                                        AttrPosAsWritten P /*=Unknown*/) {
```
A rename of 'P' to make it clear that `prettyPrintAttributes` is running in "LHS", "RHS", or "Unknown" mode would be helpful.

vgvassilev replied:
Could not understand what you mean here. What name you want for `P` here?

erichkeane replied:
I punted and hoped you had a better idea :D  Perhaps something like, `PositionToPrint` or something?  Feel free to workshop that.  Basically, I saw 'P' and thought that came from the attribute itself, so was really confused.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +241,43 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+#include "clang/Basic/AttrList.inc"
+        break;
+      default:
+        if (P == AttrPosAsWritten::Unknown || P == getAttrPosAsWritten(A, D)) {
```
I'm not sure I get the logic of 'unknown' mode here, why are we printing ALL attributes here?

Actually, when does that happen?  It seems to me that when 'P' is unknown, we should assert.  However, when `getAttrPosAsWritten` is unknown, we should probably print it 'on the left'?

vgvassilev replied:
If the mode is "Unknown" that means we want the "regular/old" behavior where we don't have a left/right preference. I realize that's probably not the best terminology I use but let me know if you have better ideas.

erichkeane replied:
Ah, hmm... So `Unknown` is kind of doing double-duty here which is concerning.  We have `Unknown` aka "I don't know where this attribute was specified!", and `Unknown` aka, "I don't want to do any additional work to specify before/after, so just print them all".

I wonder if we want two different values here, and 'assert' on the 'wrong' ones applying at the wrong time (that is, "i don't know" is invalid in this function, and "i don't care" is invalid for an attribute position-as-written).

erichkeane reviewed (commented):
Since this adds so much logic, we definitely need tests that handle when the attribute is defined in a macro, when there are multiples on each side,  when there are multiples in macros/etc.

vgvassilev commented:
> I am not sure how I feel about dropping the canPrintOnRight/printOnRight logic. We use it as a fallback when the SourceRange of a function is unreliable, otherwise we always copy and paste the code the user wrote.

Which cases do you have in mind? In theory we can always print the ones we do not know on the left. I did not see a benefit for keeping the added complexity. Can come up with an example we cannot currently support?

> 
> Regardless of that I will check for fallbacks on clang-extract once I get it to link with this branch.

If you can feed more examples that we should support that'd be great.

giulianobelinassi commented:
> > I am not sure how I feel about dropping the canPrintOnRight/printOnRight logic. We use it as a fallback when the SourceRange of a function is unreliable, otherwise we always copy and paste the code the user wrote.
> 
> Which cases do you have in mind? In theory we can always print the ones we do not know on the left. I did not see a benefit for keeping the added complexity. Can come up with an example we cannot currently support?
> 
> > Regardless of that I will check for fallbacks on clang-extract once I get it to link with this branch.
> 
> If you can feed more examples that we should support that'd be great.

Alright, so I managed to clang-extract to work on your branch and the attribute printing works as expected.
With your patch (code generated from linux-6.6/fs/ext4/super.c, `ext4_has_feature_journal` is generated by a macro):
```
__attribute__((gnu_inline))  __attribute__((unused))  __attribute__((no_instrument_function)) static inline bool ext4_has_feature_journal(struct super_block *sb) {
    return ((EXT4_SB(sb)->s_es->s_feature_compat & ((__le32)(__u32)(4))) != 0);
}
```
which seems enough for my use cases.

Notice, however, that `__attribute__(())` are now being separated by two spaces instead of one. This is mostly an aesthetic concern that anything practical.

vgvassilev commented:
> which is enough for my use cases.

Awesome to hear! Thank you!

> Notice, however, that __attribute__(()) are now being separated by two spaces instead of one. This is mostly an aesthetic concern than anything practical.

This is the annoying part I hoped to get fixed. There is one extra leading space, too... :(

giulianobelinassi commented:
> > which is enough for my use cases.
> 
> Awesome to hear! Thank you!
> 
> > Notice, however, that **attribute**(()) are now being separated by two spaces instead of one. This is mostly an aesthetic concern than anything practical.
> 
> This is the annoying part I hoped to get fixed. There is one extra leading space, too... :(

If you take a look at the diff you will see that there was code to check for an extra leading space and removing it. Perhaps you could add a boolean parameter to `prettyPrintAttributes` checking if it is the first/last attribute being print and do not issue an space there?

Or just keep the old logic and remove it from the string after it is print.

vgvassilev commented:
@erichkeane, I think I addressed your comments, can you take a look,

@giulianobelinassi, I think I found the one off space. Can you check if your case has the right amount of spaces?

vgvassilev commented on clang/test/AST/attr-print-emit.cpp:
```diff
@@ -73,3 +73,15 @@ class C {
+// CHECK: __attribute__((annotate("Annotated"))) int annotated_attr __attribute__((annotate("Annotated"))) = 0;
+
+// FIXME: We do not print the attribute as written after the type specifier.
+int ANNOTATE_ATTR annotated_attr_fixme = 0;
```
Note that attributes can appear after the type specifier, we print them on the left but not where they appeared in code unfortunately.

erichkeane replied:
Unfortunate, but not a deal breaker IMO.

vgvassilev replied:
I think we could make this work but I believe it's not worth it at that stage...

AaronBallman replied:
Hmmm, not certain if it's a deal breaker, but it sure comes close -- we support roughly 70 type attributes, including common ones like calling conventions and address spaces. So this is incremental improvement, but doesn't actually resolve the needs of either @vgvassilev or @giulianobelinassi for being able to print attributes and still compile the code. e.g., `int * [[clang::address_space(0)]] i;`

giulianobelinassi replied:
With the following input:
```
int * __attribute__((address_space(0))) i;
int * [[clang::address_space(0)]] j;
```
This prints:
```
int *__attribute__((address_space(0))) i;
int *__attribute__((address_space(0))) j
```
which seems to be semantically the same thing.

AaronBallman replied:
I would expect it to retain the original spelling of the attribute rather than swap between GNU and [[]] styles (because syntax matters greatly for appertainment with [[]] style attributes but does not [as much] for GNU-style).

Also, how does it handle monstrosities like: https://godbolt.org/z/95Pq7orde (note, implicit int behavior is definitely not a deal breaker, lol)

vgvassilev replied:
Is that something this PR breaks? I am not sure I understand if I need to change anything.

AaronBallman replied:
The PR changes behavior from one really broken state to a slightly less but still broken state. Consider:
```
[[clang::annotate("do the")]] int [[clang::annotate_type("test")]] * [[clang::annotate_type("again")]] i [[clang::annotate("and again!")]] = 0;
```
Currently we print this back as: `int *i [[clang::annotate_type(...)]] [[clang::annotate_type(...)]] [[clang::annotate("do the")]] [[clang::annotate("and again!")]] = 0;` which is all kinds of wrong. Now we print this back as: `[[clang::annotate("do the")]] int *i [[clang::annotate_type(...)]] [[clang::annotate_type(...)]] [[clang::annotate("and again!")]] = 0;` which is in better shape but still gets the type attributes wrong by moving the attribute off `int` and `*` and onto the declaration of `i`. The new code won't compile (it tries to add type attributes to a declaration, and the `...` instead of the original string literal won't compile either).

Given that printing is best-effort, perhaps this is fine because it's incremental improvement, but we're still not quite there in terms of accurate printing either.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,45 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+  SourceLocation DLoc = D->getLocation();
+  const ASTContext &C = D->getASTContext();
+  if (ALoc.isInvalid() || DLoc.isInvalid())
+    return DeclPrinter::AttrPosAsWritten::Default;
```
So for this one, I thought `Unknown` made sense here.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,45 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+#include "clang/Basic/AttrList.inc"
+        break;
+      default:
+        if (Pos == AttrPosAsWritten::Default || Pos == getPosAsWritten(A, D)) {
```
Even by the new name, `Default` is doing double-duty, and that is confusing to me.  See my suggestion about adding asserts + a 2nd 'position'.

vgvassilev replied:
Ok, I addressed this.

erichkeane commented on clang/test/OpenMP/assumes_template_print.cpp:
```diff
@@ -17,23 +17,23 @@ template <typename T>
   int a;
 // CHECK: template <typename T> struct S {
-// CHECK:     __attribute__((assume("ompx_global_assumption"))) void foo()     {
+// CHECK:     void foo() __attribute__((assume("ompx_global_assumption")))     {
```
What causes these to swap sides?

vgvassilev replied:
I think that attribute is added by `#pragma omp parallel` which appears after the definition of the function in terms of source locations. I thought we should probably filter out attributes that came from pragmas, however, it seems that we do not retain that information in the attribute and it says "gnu" rather than "pragma"...

erichkeane replied:
Is it perhaps set as "isImplicit"?  I wonder if we could filter out based on that?

vgvassilev replied:
It is not implicit - we don't print them. I looked and could not find anything that could distinguish the ones that come from pragmas. I briefly looked to see if we can add this in but seems some more work to be able to ask `::isOMPAttr` in tablegen. Generally that's a good improvement but it has a different timeline to implement than that PR which is meant to be a hot fix...

erichkeane replied:
@alexey-bataev : Comments/concerns?

erichkeane commented on clang/test/OpenMP/declare_simd_ast_print.cpp:
```diff
@@ -60,11 +60,11 @@ void h(int *hp, int *hp2, int *hq, int *lin)
   // CHECK-NEXT: }
   #pragma omp declare simd uniform(this, a) linear(val(b): a)
-  __attribute__((cold)) int add(int a, int b) { return a + b; }
+  int add(int a, int b) __attribute__((cold)) { return a + b; }
```
Why did you change where the attribute was placed here?  is there a problem leaving it on the left?

vgvassilev replied:
Same reason I mentioned above. It's introduced by a pragma whose source location we take.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,47 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
 
   if (D->hasAttrs()) {
-    AttrVec &Attrs = D->getAttrs();
+    const AttrVec &Attrs = D->getAttrs();
```
nit: I'd like an assert here to make sure `Pos != Unknown` (or at the beginning of hte function).

vgvassilev replied:
I added an assert in the if-stmt. Is that what you meant?

erichkeane replied:
I see that, it does what I want, yes.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,47 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+        break;
+      default:
+        AttrPosAsWritten APos = getPosAsWritten(A, D);
+        assert(APos != AttrPosAsWritten::Unknown && "Implicit attribute!");
```
Isn't this the inverse?  `APos` CAN be `Unknown`, it is `Default` that we need to make sure it isn't.

vgvassilev replied:
APos cannot be unknown because the loop continues `if (A->isInherited() || A->isImplicit())`.

erichkeane replied:
IS that the only case where either the attribute or declaration don't have a valid source location?  I find myself wondering about future attributes with a deduction guide/etc (or some other declaration that we represent as an AST node that gets generated out of thin air, that also might get an attribute from a spelled thing).

erichkeane replied:
Also, probably should make sure APos isn't 'default' either via assert.

AaronBallman replied:
I think if `APos` is unknown, then there's a bug elsewhere. You can create attributes manually without using `CreateImplicit()` and so it's possible to add attributes with invalid source locations. For example, I spotted:
https://github.com/llvm/llvm-project/blob/ea88bb16485a34db58333637d8558ff57e6a1329/clang/lib/Sema/SemaAPINotes.cpp#L586 and https://github.com/llvm/llvm-project/blob/ea88bb16485a34db58333637d8558ff57e6a1329/clang/lib/Sema/SemaDecl.cpp#L6884 which will probably trigger this assertion.

So I think the assertion is reasonable but should probably come with a comment explaining why it might be triggered and what to do if it is (switch things to using `CreateImplicit()`).

vgvassilev replied:
Comment added.

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,47 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+        assert(APos != AttrPosAsWritten::Unknown && "Implicit attribute!");
+        if (Pos == AttrPosAsWritten::Default || Pos == APos) {
+          if (Pos != AttrPosAsWritten::Left)
+            Out << " ";
```
```suggestion
            Out << ' ';
```

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,47 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+            Out << " ";
+          A->printPretty(Out, Policy);
+          if (Pos == AttrPosAsWritten::Left)
+            Out << " ";
```
```suggestion
            Out << ' ';
```

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,47 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+      default:
+        AttrPosAsWritten APos = getPosAsWritten(A, D);
+        assert(APos != AttrPosAsWritten::Unknown && "Implicit attribute!");
+        if (Pos == AttrPosAsWritten::Default || Pos == APos) {
```
In the case we are printing `Left` and `Right`, where do attributes we can't figure out based on location get printed?  That is, where do `Unknown` location attributes get printed?

vgvassilev replied:
If we have not specified `Left` or `Right` it prints them in their default position which seems predominantly on the right. In theory, we can rework all calls to `prettyPrintAttributes` and remove the need of `Default` because we will have:
```
printing_logic;
printAttributes(left);
more_printing_logic;
printAttributes(right);
```

That's doable but seemed outside of the scope of this PR since I am hoping to get it merged asap...

giulianobelinassi commented:
> @erichkeane, I think I addressed your comments, can you take a look,
> 
> @giulianobelinassi, I think I found the one off space. Can you check if your case has the right amount of spaces?

I don't see those extra spaces on the current patch anymore. Nice work!

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,50 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+        AttrPosAsWritten APos = getPosAsWritten(A, D);
+        // Might trigger on programatically created attributes with no source
+        // locations.
+        assert(APos != AttrPosAsWritten::Unknown && "Implicit attribute!");
```
```suggestion
        assert(APos != AttrPosAsWritten::Unknown &&  "Invalid source location for attribute or decl.");
        assert(APos !=AttrPosAsWritten::Default && "Default not a valid for an attribute location");
```

erichkeane commented on clang/lib/AST/DeclPrinter.cpp:
```diff
@@ -250,87 +238,50 @@ raw_ostream& DeclPrinter::Indent(unsigned Indentation) {
+        break;
+      default:
+        AttrPosAsWritten APos = getPosAsWritten(A, D);
+        // Might trigger on programatically created attributes with no source
```
```suggestion
        // Might trigger on programatically created attributes or declarations with no source
```

vgvassilev commented:
@erichkeane, thank you. What's the process of including this in the next release?

vgvassilev commented:
> > @erichkeane, thank you. What's the process of including this in the next release?
> 
> After CI is complete, you can click "Squash and Merge" below (if you cannot, let us know and someone can do it for you), and it'll be included in the 19.1 release this summer.

I have commit access. I want this to be part of the 18.x releases as that breaks our downstream clients.

erichkeane commented:
> > > @erichkeane, thank you. What's the process of including this in the next release?
> > 
> > 
> > After CI is complete, you can click "Squash and Merge" below (if you cannot, let us know and someone can do it for you), and it'll be included in the 19.1 release this summer.
> 
> I have commit access. I want this to be part of the 18.x releases as that breaks our downstream clients.

Ah, hmm... I am not sure this qualifies for inclusion in the current release branch.  Perhaps @AaronBallman can comment here.

AaronBallman commented:
> > > > @erichkeane, thank you. What's the process of including this in the next release?
> > > 
> > > 
> > > After CI is complete, you can click "Squash and Merge" below (if you cannot, let us know and someone can do it for you), and it'll be included in the 19.1 release this summer.
> > 
> > 
> > I have commit access. I want this to be part of the 18.x releases as that breaks our downstream clients.
> 
> Ah, hmm... I am not sure this qualifies for inclusion in the current release branch. Perhaps @AaronBallman can comment here.

I'm not particularly comfortable putting this into 18.x; we should only be pushing very safe fixes to regressions and this one doesn't really qualify. It's debatable whether it's a regression (it sort of is, sort of isn't), but the changes are also relatively involved. I'd feel more comfortable with this in 19.x so we have more time to find and fix remaining edge cases.

vgvassilev commented:
Ok. Fair enough.

giulianobelinassi commented:
> > > > > @erichkeane, thank you. What's the process of including this in the next release?
> > > > 
> > > > 
> > > > After CI is complete, you can click "Squash and Merge" below (if you cannot, let us know and someone can do it for you), and it'll be included in the 19.1 release this summer.
> > > 
> > > 
> > > I have commit access. I want this to be part of the 18.x releases as that breaks our downstream clients.
> > 
> > 
> > Ah, hmm... I am not sure this qualifies for inclusion in the current release branch. Perhaps @AaronBallman can comment here.
> 
> I'm not particularly comfortable putting this into 18.x; we should only be pushing very safe fixes to regressions and this one doesn't really qualify. It's debatable whether it's a regression (it sort of is, sort of isn't), but the changes are also relatively involved. I'd feel more comfortable with this in 19.x so we have more time to find and fix remaining edge cases.

How about this single-line fix? It would fix the override regression for now.
```
diff --git a/clang/include/clang/Basic/Attr.td b/clang/include/clang/Basic/Attr.td
index 80e607525a0a..a39288464040 100644
--- a/clang/include/clang/Basic/Attr.td
+++ b/clang/include/clang/Basic/Attr.td
@@ -2511,6 +2511,7 @@ def Overloadable : Attr {
 }
 
 def Override : InheritableAttr {
+  let CanPrintOnLeft = 0;
   let Spellings = [CustomKeyword<"override">];
   let SemaHandler = 0;
   // Omitted from docs, since this is language syntax, not an attribute, as far
   ```

vgvassilev commented:
Great idea! That’d make sense to me.

giulianobelinassi commented:
We would like this fixed in 18.1 as well. We are expanding to support C++ and we will hit this bug at some point.

vgvassilev commented:
Maybe you can open a PR against the branch?

giulianobelinassi commented:
> Maybe you can open a PR against the branch?

Sorry, but I can only do this tomorrow. Feel free to open the PR if you need it immediately.

> And `final` as well as `override`? (This is why I'm not convinced we should be backporting anything -- the problem is with printing in general and will crop up in various places, so we're not really fixing a regression so much as playing whack-a-mole with a few cases.)

Why not? `override` and perhaps `final` output is broken so I'd say we fix this on LLVM-18. It is better than leave it broken. I don't think this will break many test cases once this bug got undetected.

kimgr commented:
A note from left field: I think this PR broke the IWYU test suite. We use `TemplateDecl::print` + some postprocessing to generate template forward-declarations. We're seeing this diff now:
```diff
-template <typename T> class FinalTemplate;
+template <typename T> class  FinalTemplate;
```
So a spurious extra space.

We would be grateful if this _didn't_ make the 18.x branches, because we've already released an 18.x-compatible version of IWYU, and it would be a shame if that release had a failing test suite when built against later 18.x Clang.

I'll work around this on our latest mainline.

erichkeane commented:
> A note from left field: I think this PR broke the IWYU test suite. We use `TemplateDecl::print` + some postprocessing to generate template forward-declarations. We're seeing this diff now:
> 
> ```diff
> -template <typename T> class FinalTemplate;
> +template <typename T> class  FinalTemplate;
> ```
> 
> So a spurious extra space.
> 
> We would be grateful if this _didn't_ make the 18.x branches, because we've already released an 18.x-compatible version of IWYU, and it would be a shame if that release had a failing test suite when built against later 18.x Clang.
> 
> I'll work around this on our latest mainline.

Ooof, I think that DOES decide it for me.  @vgvassilev : can you look into that extra space to see if you can track it down?

kimgr commented:
@erichkeane Thanks! Let me just see if I can bisect it to this commit; I don't have any evidence yet, this PR was just the only significant change to `DeclPrinter.cpp` in the past few days. I need a little while to rebuild my local Clang tree with and without the patch.

kimgr commented:
I can confirm that the double space comes from this PR;
```diff
diff --git a/clang/unittests/AST/DeclPrinterTest.cpp b/clang/unittests/AST/DeclPrinterTest.cpp
index c24e442621c9..c2d02e74a62c 100644
--- a/clang/unittests/AST/DeclPrinterTest.cpp
+++ b/clang/unittests/AST/DeclPrinterTest.cpp
@@ -1555,3 +1555,11 @@ TEST(DeclPrinter, VarDeclWithInitializer) {
       PrintedDeclCXX17Matches("void foo() {int arr[42]; for(int a : arr);}",
                               namedDecl(hasName("a")).bind("id"), "int a"));
 }
+
+TEST(DeclPrinter, TestTemplateFinal) {
+  ASSERT_TRUE(PrintedDeclCXX11Matches(
+      "template<typename T>\n"
+      "class FinalTemplate final {};",
+      classTemplateDecl(hasName("FinalTemplate")).bind("id"),
+      "template <typename T> class final FinalTemplate final {}"));
+}
```
fails with:
```
.../llvm-project/clang/unittests/AST/DeclPrinterTest.cpp:1560: [...]
  Expected "template <typename T> class final FinalTemplate final {}"
  got           "template <typename T> class  final FinalTemplate final {}")
```
(edited so it's easier to see the diff).

It passes after I revert the `Rework attributes` commits:

* 9391ff8c86007562d40c240ea082b7c0cbf35947
* 62e92573d28d62ab7e6438ac34d513b07c51ce09
* a30662fc2acdd73ca1a9217716299a4676999fb4

Again, from my point of view, this "bug" is fine on mainline, we can work around it. But it would be nice if this patch was not backported to 18 as it breaks our corresponding release.

kimgr commented:
(obtw, the double `final` is unrelated, it's tracked here: https://github.com/llvm/llvm-project/issues/56517)

vgvassilev commented:
> I can confirm that the double space comes from this PR;
> 
> ```diff
> diff --git a/clang/unittests/AST/DeclPrinterTest.cpp b/clang/unittests/AST/DeclPrinterTest.cpp
> index c24e442621c9..c2d02e74a62c 100644
> --- a/clang/unittests/AST/DeclPrinterTest.cpp
> +++ b/clang/unittests/AST/DeclPrinterTest.cpp
> @@ -1555,3 +1555,11 @@ TEST(DeclPrinter, VarDeclWithInitializer) {
>        PrintedDeclCXX17Matches("void foo() {int arr[42]; for(int a : arr);}",
>                                namedDecl(hasName("a")).bind("id"), "int a"));
>  }
> +
> +TEST(DeclPrinter, TestTemplateFinal) {
> +  ASSERT_TRUE(PrintedDeclCXX11Matches(
> +      "template<typename T>\n"
> +      "class FinalTemplate final {};",
> +      classTemplateDecl(hasName("FinalTemplate")).bind("id"),
> +      "template <typename T> class final FinalTemplate final {}"));
> +}
> ```
> 
> fails with:
> 
> ```
> .../llvm-project/clang/unittests/AST/DeclPrinterTest.cpp:1560: [...]
>   Expected "template <typename T> class final FinalTemplate final {}"
>   got      "template <typename T> class  final FinalTemplate final {}")
> ```
> 
> (edited so it's easier to see the diff).
> 
> It passes after I revert the `Rework attributes` commits:
> 
>     * [9391ff8](https://github.com/llvm/llvm-project/commit/9391ff8c86007562d40c240ea082b7c0cbf35947)
> 
>     * [62e9257](https://github.com/llvm/llvm-project/commit/62e92573d28d62ab7e6438ac34d513b07c51ce09)
> 
>     * [a30662f](https://github.com/llvm/llvm-project/commit/a30662fc2acdd73ca1a9217716299a4676999fb4)
> 
> 
> Again, from my point of view, this "bug" is fine on mainline, we can work around it. But it would be nice if this patch was not backported to 18 as it breaks our corresponding release.

@kimgr, than you for the report. I can reproduce the double space. However, do you expect that test of yours to be valid C++? As written it seems not.

kimgr commented:
@vgvassilev I did expect the input to be valid, yes:
```
template<typename T>
class FinalTemplate final {};
```
Is it not?

vgvassilev commented:
> @vgvassilev I did expect the input to be valid, yes:
> 
> ```
> template<typename T>
> class FinalTemplate final {};
> ```
> 
> Is it not?

The snippet as visualized in github seems to have one too many `final`s: `template <typename T> class final FinalTemplate final {}`

kimgr commented:
Ah, that's the expected output -- I can't do anything about that :). See https://github.com/llvm/llvm-project/issues/56517.

vgvassilev commented:
> Ah, that's the expected output -- I can't do anything about that :). See #56517.

I believe this should fix it: https://github.com/llvm/llvm-project/pull/88600 Can you test?
